src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdateService.java

```diff
@@ -0,0 +1,123 @@
+        return directDebitEventService.registerMandateCancelledEventFor(newMandate);
+    }
+    public DirectDebitEvent awaitingDirectDebitDetailsFor(Mandate mandate) {
+        return directDebitEventService.registerAwaitingDirectDebitDetailsEventFor(mandate);
+    }
+
+    public DirectDebitEvent mandateActiveFor(Mandate mandate) {
+        updateStateFor(mandate, MANDATE_ACTIVE);
+        return directDebitEventService.registerMandateActiveEventFor(mandate);
+    }
+
+    public DirectDebitEvent mandateExpiredFor(Mandate mandate) {
+        Mandate updatedMandate = updateStateFor(mandate, SupportedEvent.MANDATE_EXPIRED_BY_SYSTEM);
+        return directDebitEventService.registerMandateExpiredEventFor(updatedMandate);
+    }
+    public DirectDebitEvent mandateFailedFor(Mandate mandate) {
+        Mandate newMandate = updateStateFor(mandate, MANDATE_FAILED);
+        userNotificationService.sendMandateFailedEmailFor(newMandate);
+        return directDebitEventService.registerMandateFailedEventFor(newMandate);
+    }
+
+    public DirectDebitEvent mandateCancelledFor(Mandate mandate) {
+        Mandate newMandate = updateStateFor(mandate, MANDATE_CANCELLED);
+        userNotificationService.sendMandateCancelledEmailFor(newMandate);
+        return directDebitEventService.registerMandateCancelledEventFor(newMandate);
+    }
+
+    public DirectDebitEvent mandatePendingFor(Mandate mandate) {
+        Mandate newMandate = updateStateFor(mandate, MANDATE_PENDING);
+        return directDebitEventService.registerMandatePendingEventFor(newMandate);
+    }
+    public Mandate receiveDirectDebitDetailsFor(Mandate mandate) {
+        directDebitEventService.registerDirectDebitReceivedEventFor(mandate);
+        return mandate;
+    }
+
+    public Mandate payerCreatedFor(Mandate mandate) {
+        directDebitEventService.registerPayerCreatedEventFor(mandate);
+        return mandate;
+    }
+
+    public DirectDebitEvent payerEditedFor(Mandate mandate) {
+        return directDebitEventService.registerPayerEditedEventFor(mandate);
+    }
+    
+    public Mandate tokenExchangedFor(Mandate mandate) {
+        Mandate newMandate = updateStateFor(mandate, TOKEN_EXCHANGED);
+        directDebitEventService.registerTokenExchangedEventFor(newMandate);
+        return newMandate;
+    }
+    private Mandate updateStateFor(Mandate mandate, DirectDebitEvent.SupportedEvent event) {
+        MandateState newState = MandateStatesGraph.getStates().getNextStateForEvent(mandate.getState(),
+                event);
+        mandateDao.updateState(mandate.getId(), newState);
+        LOGGER.info("Updating mandate {} - from {} to {}",
+                mandate.getExternalId(),
+                mandate.getState(),
+                newState);
+        mandate.setState(newState);
+        return mandate;
+    }
+    
+    void canUpdateStateFor(Mandate mandate, DirectDebitEvent.SupportedEvent event) {
```

Comment:
I know I wrote this method, but it is a bit weird - I would prefer it to return a Boolean rather than throwing

Reply:
but again, not a big deal

Reply:
hahaha...not sure about that, i think it fits well with the exception mapper pattern we are using. It's a bit of a guard. If it returns a boolean then we just move the logic (if it can't update, throw) one level above

Reply:
jolly good